# Numerical Linear Algebra & Parallel Computing

### KIDDI MOHAMED

###### Nouredine Ouhaddou ◾ Mohamed Jalal Maaouni ◾ Ibrahim El Mountasser

### Complexity Analysis

#Introduction
1) Describe solution 1
2) Describe solution 2
3) Run the two programs for different values of n and measure which algorithm is faster.
4) Calculate the number of operations executed by each of the programs for different
values of n and generalize for any n.

### First Solution

In [178]:
def count_divisors1(n):
    count=0
    d=1
    while d<=n:
        if n%d==0:
            count+=1
        d+=1
    return count

### Second solution

In [179]:
def count_divisors2(n):
    count=0
    d=1
    while d*d<=n:
        if n%d==0:
            count+=1 if n/d==d else 2
        d+=1
    return count

### Other Solution

In [180]:
def count_divisors3(n):
    count=1
    d=1
    while d <= n/2:
        if n%d==0:
            count+=1
        d+=1
    return count

###### 1) Describe solution 1

For the first solution we iterate in the while loop for n times and for each time we check if the value of n%d==0 , (ie check if d devide n or not) if yes we increment the values of 'count' then as a  result we will find the number of divisors of a given n 

###### 2) Describe solution 2

for the second solution we iterate just for sqrt(n) and based on a rule that if d is a divisor of n, then n/d is also a divisor of n. \
Therefore, instead of iterating through all integers from 1 to n, it only iterates through all integers from 1 to sqrt(n) and check if they devide n in this case we will increment by 2 , but if n/d to d then we will just increment by 1. \
and as the first function  , our function retun the numbers of divisors for a given n .

###### 3) Run the two programs for different values of n and measure which algorithm is faster. 

In [190]:
a= %time count_divisors1(1000)

CPU times: user 65 µs, sys: 0 ns, total: 65 µs
Wall time: 66 µs


In [182]:
%time count_divisors2(1000)

CPU times: user 6 µs, sys: 0 ns, total: 6 µs
Wall time: 7.39 µs


16

In [192]:
%time count_divisors3(1000)

CPU times: user 42 µs, sys: 0 ns, total: 42 µs
Wall time: 43.9 µs


16

For n=1000
The faster solution is the second one  *Wall time: 7.39 µs*, follow by the new solution *Wall time: 43.9 µs* and finally  the first one *Wall time: 66 µs*


In [184]:
%time count_divisors1(100000)

CPU times: user 5.24 ms, sys: 0 ns, total: 5.24 ms
Wall time: 5.3 ms


36

In [185]:
%time count_divisors2(100000)

CPU times: user 29 µs, sys: 0 ns, total: 29 µs
Wall time: 31 µs


36

In [186]:
%time count_divisors3(100000)

CPU times: user 3.67 ms, sys: 0 ns, total: 3.67 ms
Wall time: 3.7 ms


36

For n=100000
The faster solution is the second one  *Wall time: 31 µs*, follow by the new solution *Wall time: 3.7 ms* and finally  the first one *Wall time: 5.3 ms*


In [193]:
%time count_divisors1(100000000)

CPU times: user 4.83 s, sys: 0 ns, total: 4.83 s
Wall time: 4.84 s


81

In [188]:
%time count_divisors2(100000000)

CPU times: user 758 µs, sys: 0 ns, total: 758 µs
Wall time: 759 µs


81

In [189]:
%time count_divisors3(100000000)

CPU times: user 3.31 s, sys: 0 ns, total: 3.31 s
Wall time: 3.31 s


81

For n=100000000
The faster solution is the second one  *Wall time: 759 µs*, follow by the new solution *Wall time: 3.31 s* and finally  the first one *Wall time: 4.84 s*


Then we notice that the second solution is the faster one , and there is a big difference between the second one and the others .

###### 4) Calculate the number of opera.ons executed by each of the programs for different values of n and generalize for any n. 

n =1000 \
For the first solution : \
there is two operation of declaration (count and d )also the loop while execute 1000 times in each time there is one modulo operation  and one addition(d+=1) then 2000, and we will add the addition (count+=1) count times , and in n=1000 count = 16 \
then the number of operations in 2018
n =100000 \
the same logic we will find number of operation is 200038 \
n =100000000 \
the same logic we will find number of operation is 200000083 \
then 
number of operation= 2\*n+2+count.

As we notice that the value of count is negligible then we can say that : \
number of operation is equivalent to 2*n then we can conclude that the complexity of the first solution is O(n).



n =1000 \
For the second solution : \
there is two operation of initialization of count and d ,and a while loop that iterates sqrt(1000) times, and in each iteration there is n%d==0 the total number of operations executed by this function for n=1000 is  equivalent to 2sqrt(1000) + 2 

n =100000 \
the same logic we will find number of operation is equivalent to  2sqrt(100000) + 2 \ 

in this solution we can t calculate the exacte number of operation because we cannot use the variable count as we did before , because it didn t give us the number of time that the two operation 'n/d==d' and  'count+=1 or 2 ' happends . \
then we just estimate the values and trying to negligate the other values 

In general The number of operation is close to 2sqrt(n)+2 therefore the complexity of the second solution is O(sqrt(n)) 


#### Big-O notation
###### T(n)=3 *n^3+2*n^2+ 1/2*n+7, prove that T(n)=O(n^3)

if eventually for all sufficiently large n, T(n) is bounded above by 3*n^3   We can see that as n grows, the dominant term becomes n^3. 
then the complexity of T(n) is n^3

###### Prove that for every k>=1, n^k is not O(n^(k-1))

To prove it ,we will use a proof by contradiction .
Assume that for some k>=1, n^k is O(n^(k-1)) \ 
then by definition of big O notation, there exist constants c and n0 such that n^k <= c*n^(k-1) for all n >= n0
if we divide the both side by n^(k-1), we get n <= c for all n >= n0
This is a contradiction.
ince for any value of c, there will always exist an n such that n > c, and thus the inequality n <= c cannot hold for all n >= n0.
Therefore, we can conclude that for every k>=1, n^k is not O(n^(k-1))

### Merge sort

In [1]:
def merge(A, B):
    n = len(A)
    m = len(B)
    i = 0
    j = 0
    C = []
    while i < n and j < m:
        if A[i] <= B[j]:
            C.append(A[i])
            i += 1
        else:
            C.append(B[j])
            j += 1
    while i < n:
        C.append(A[i])
        i += 1
    while j < m:
        C.append(B[j])
        j += 1
    return C


In [6]:
A=[1,3,5,7,9]
B=[2,4,6,8,10,12]
C=merge(A,B)
C

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12]

##### 2) Analyse the complexity of your function using Big-O notation.

The time complexity of the first code is O(nm) because the outer loop iterates through each element of array A, and the inner loop iterates through each element of array B, resulting in n times m iterations in total if A has n elements and B has m elements.
that if we assume that the A and B Are Already sorted 

### The master method

##### 1) Using the master method analyse the complexity of merge sort.

The rule that the master method follows is expressed as T(n) = a * T(n/b) + f(n), where T(n) represents the time complexity of an algorithm for a given input size n. Here, a is the number of subproblems generated at each recursive level, n/b is the size of the subproblem, and f(n) is the cost of combining the solutions and dividing the problems.

In this particular case, we have T(n) = 2 * T(n/2) + O(n) (where f(n) = O(n¹) and d = 1), with a = 2 and b = 2. We can then check the second condition of the master method, which is that log_b(a) = log_2(2) = 1. This means that the time complexity of the algorithm is O(n^d log(n)), where d = 1. Therefore, the overall time complexity is O(nlog(n)).

#### 2) Using the master method analyse the complexity of binary search

 the time complexity of the algorithm is expressed as T(n) = T(n/2) + O(1), where f(n) is equal to O(n⁰) since d = 0. Here, a = 1 and b = 2.
we apply the second condition of the master method, which states that log_b(a) = log_2(1) = 0. This implies that the time complexity of the algorithm is O(n^d log(n)), with a = 1 and b = 2. In this case, the complexity is simply O(log(n)).

#### Bonus

### Matrix multiplication


#### 1) Write a func.on using python3 that mul.ply two matrices A,B (without the use of numpy or any external library)

In [26]:
def multi(A,B):
    m=len(A)
    n=len(A[0])
    o=len(B)
    p=len(B[0])
    c = [[0 for row in range(p)]
        for col in range(m)]
    
    if n!=o :
        return 0
    
    for i in range(m):
        for j in range(p):
            for k in range(n):
                c[i][j] += A[i][k] * B[k][j]
    return c

In [27]:
A = [[1, 4, 5], 
    [-5, 8, 9]]
B = [[1, 4, 5], 
    [-5, 8, 9],
      [1,1,1]]
multi(A,B)

[[-14, 41, 46], [-36, 53, 56]]

#### 2) What’s the complexity of your algorithm (using big-O notation)?

the complexity of our code is O(mpn), and in the simple case where our matrix are square the complexity become O(n^3)

#### 3) Write the same func.on in C. (bonus)

In [28]:
#include <stdio.h>
#include <stdlib.h>

int** multi(int **A, int **B, int m, int n, int o, int p) {
    int i, j, k;
    int **C = (int **)malloc(m * sizeof(int *));
    for (i = 0; i < m; i++)
        C[i] = (int *)calloc(p, sizeof(int));

    if (n != o)
        return 0;

    for (i = 0; i < m; i++) {
        for (j = 0; j < p; j++) {
            for (k = 0; k < n; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return C;
}


SyntaxError: invalid syntax. Perhaps you forgot a comma? (657256586.py, line 4)

4) Optimize this multiplication and describe each step of your optimisation.

1) Use list comprehension to initialize the result matrix C:
Instead of using nested loops to initialize each element of the result matrix C, we can use a list comprehension to initialize it all at once. This is faster because list comprehensions are optimized in Python and are executed more quickly than nested loops.

2)we can also use block matrix multiplication:
By breaking down the matrices into smaller blocks, we can improve the efficiency of the algorithm by maximizing cache utilization and minimizing the amount of data accessed from memory.

In [41]:
def multi(A, B):
    m = len(A)
    n = len(A[0])
    o = len(B)
    p = len(B[0])

    if n != o:
        return 0

    # Use a list comprehension to initialize the result matrix C
    # This is faster than using a nested loop to initialize each element
    C = [[0] * p for _ in range(m)]
    
    for i in range(m):
        for j in range(p):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]
    return C

In [42]:
multi(A,B)

[[-14, 41, 46], [-36, 53, 56]]